src/serverlist.cpp

```diff
@@ -631,6 +635,10 @@ void CServerListManager::Remove ( const CHostAddress& InetAddr )
  - SERVER external to DIRECTORY (list entry external IP same LAN)
    - CLIENT internal to SERVER (CLM     same IP as list entry external IP): use "internal" fields (local LAN address)
    - CLIENT external to SERVER (CLM not same IP as list entry external IP): use "external" fields (public IP address from protocol)
+
+ Finally, when retrieving the entry for the directory itself life is more complicated still.  It's easiest just to return "0"
+ and hope the IP and UDP fill in the gaps.
```

Comment:
That sounds a bit unfounded. Do we have better facts than hoping? ;)

Reply:
:)  Well, it side-step putting an actual definite value into the field.  It puts zero in and lets IP/UDP decide, _I think_.  Either that, or I'm missing something in `src/protocol.cpp`.

Reply:
Hm, I don't think IP/UDP can or should decide that. I think the protocol layer could decide, but [I don't see special casing for zero HostAddress fields](https://github.com/jamulussoftware/jamulus/blob/05f385c8019269369fb7fd5f0f6b038dd49ab84c/src/protocol.cpp#L2036-L2042). The [receiving side does not show special-casing either](https://github.com/jamulussoftware/jamulus/blob/master/src/protocol.cpp#L2080-L2081).

It seems like it is once again the **GUI** which contains relevant logic (and shouldn't...):
https://github.com/jamulussoftware/jamulus/blob/05f385c8019269369fb7fd5f0f6b038dd49ab84c/src/connectdlg.cpp#L346-L354

After reading the logic, that the client always special-cases index 0, I'm no longer sure why the server-set address is relevant at all.

Did someone confirm that it affects Jamulus?
Could it be that it's "just" Jamulus Explorer which takes what the directory sends?

(I'm not saying we shouldn't fix it, as it is obviously an unintended change of the previous release; but it would be less critical if it didn't affect vanilla Jamulus usage)

Reply:
Jamulus Explorer works different: It does not special-case index 0, instead [it lets 0 IP addresses default to the directory address](https://github.com/softins/jamulus-php/blob/d3b295a594ad35a3891033db31fcd2e77dbde905/servers.php#L753-L756). Therefore, this is consistent with how 3.9.0 behaves and should be remedied with this PR.

Reply:
Hm...  I don't like it that it's in `src/connectdlg.cpp` -- _no_ decision making processing should be done in the dialogue layer, only display/formatting processing.  This should either be in the protocol layer (when receiving the list) or in the client layer (when returning the retrieved list).

Admittedly, the server list is only used by the connect dialogue...